File: back_end/models/productModel.js
const mongoose = require('mongoose');
const { toThaiSlug } = require('../utils/toThaiSlug');

const productSchema = new mongoose.Schema({
    name: {
        type: String,
        required: [true, 'A product must have a name'],
        trim: true,
        maxlength: [40, 'A product name must have less or equal then 40 characters'],
        // minlength: [10, 'A product name must have more or equal then 10 characters']
    },
    imageCover: {
        type: String,
        // default: 'default.jpg'
        default: 'default.jpeg',
        required: [true, 'A product must have a cover image']
    },
    images: {
        type: Array,
        default: []
    },
    slug: String,
    description: {
        type: String,
        trim: true,
        required: [true, 'A product must have a description'],
        maxlength: [200, 'A product description must have less or equal then 200 characters'],
    },
    category: {
        type: String,
        enum: ["processed", "otop", "medicinalPlant", "driedGood"],
        required: [true, 'A product must have a category']
    },
    quantity: {
        type: Number,
        required: [true, 'A product must have a quantity']
    },
    price: {
        type: Number,
        required: [true, 'A product must have a price'],
        min: [0, 'Price must be above 0']
    }, discount: {
        type: Number,
        default: 0,
        min: [0, 'Discount must be above 0'],
        max: [100, 'Discount must be below 100']
    },
    sold: {
        type: Number,
        default: 0,
        min: [0, 'Sold must be above 0']
    },

    // shipping: {
    //     type: String,
    //     enum: ['Yes', 'No']
    // },
    // color: {
    //     type: String,
    //     enum: ['Black', 'Brown', 'Silver', 'White', 'Blue']
    // },
    // brand: {
    //     type: String,
    //     enum: ['Apple', 'Samsung', 'Microsoft', 'Lenovo', 'ASUS']
    // },
    // ratings: {
    //     type: Number,
    //     default: 0,
    //     min: [0, 'Rating must be above 0'],
    //     max: [5, 'Rating must be below 5'],
    // },
    ratingsAverage: {
        type: Number,
        default: 5,
        min: [1, 'Rating must be above 1.0'],
        max: [5, 'Rating must be below 5.0'],
        set: val => Math.round(val * 10) / 10 // 4.666666, 46.6666, 47, 4.7
    },
    ratingsQuantity: {
        type: Number,
        default: 0
    },
    shop: {
        type: mongoose.Schema.ObjectId,
        ref: 'Shop',
        required: [true, 'A product must belong to a shop']
    },
    // reviews: [
    //     {
    //         type: mongoose.Schema.Types.ObjectId,
    //         ref: 'Review'
    //     }
    // ],


},
    {
        timestamps: true,
        toJSON: { virtuals: true },
        toObject: { virtuals: true }
    },
    this.collection = 'products',
);
productSchema.index({ price: 1, ratingsAverage: -1 });
productSchema.index({ slug: 1 });

// Convert name to slug
// Craete middleware
productSchema.pre('save', function (next) {
    this.slug = toThaiSlug(this.name);
    next();
});


// Reverse populate with virtuals
productSchema.virtual('reviews', {
    ref: 'Review',
    foreignField: 'product', // field in the Review model
    localField: '_id' // field in the Product model
});

// Reverse populate with virtuals
productSchema.virtual('shops', {
    ref: 'Shop',
    foreignField: 'products', // field in the Review model
    localField: '_id' // field in the Product model
});



module.exports = mongoose.model('Product', productSchema);

File: front_end/lib/src/models/product_model.dart
import 'shop_model.dart';

class ProductDetail {
  final String id;
  final String name;
  final List<String> images;
  final String description;
  final String category;
  final int quantity;
  final double price;
  final double discount;
  final int sold;
  final double ratingsAverage;
  final int ratingsQuantity;
  final Shop shop;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String slug;
  final String imageCover;

  ProductDetail({
    required this.id,
    required this.name,
    required this.images,
    required this.description,
    required this.category,
    required this.quantity,
    required this.price,
    required this.discount,
    required this.sold,
    required this.ratingsAverage,
    required this.ratingsQuantity,
    required this.shop,
    required this.createdAt,
    required this.updatedAt,
    required this.slug,
    required this.imageCover,
  });

  factory ProductDetail.empty() {
    return ProductDetail(
      id: '',
      name: '',
      images: [],
      description: '',
      category: '',
      quantity: 0,
      price: 0.0,
      discount: 0.0,
      sold: 0,
      ratingsAverage: 0.0,
      ratingsQuantity: 0,
      shop: Shop.empty(),
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
      slug: '',
      imageCover: '',
    );
  }

  factory ProductDetail.fromJson(Map<String, dynamic> json) {
    return ProductDetail(
      id: json['_id'],
      name: json['name'],
      images: List<String>.from(json['images']),
      description: json['description'],
      category: json['category'],
      quantity: json['quantity'],
      price: json['price'].toDouble(),
      discount: json['discount'].toDouble(),
      sold: json['sold'],
      ratingsAverage: json['ratingsAverage'].toDouble(),
      ratingsQuantity: json['ratingsQuantity'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      slug: json['slug'],
      imageCover: json['imageCover'],
      shop: Shop.fromJson(
        json['shop'] ??
            {
              // 'name': '',
              // 'location': '',
            },
      ),
    );
  }
}

class Product {
  final String name;
  final double price;
  final int discount;
  final int sold;
  final String imageCover;
  final String id;
  final int quantity;

  Product({
    required this.name,
    required this.price,
    required this.discount,
    required this.sold,
    this.imageCover = '',
    required this.id,
    required this.quantity,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      name: json['name'] ?? '',
      price: json['price'].toDouble() ?? 0.0,
      discount: json['discount'].toInt() ?? 0,
      sold: json['sold'] ?? 0,
      imageCover: json['imageCover'] ?? '',
      id: json['_id'] ?? '',
      quantity: json['quantity'] ?? 0,
    );
  }
}

class CheckoutProduct {
  final String shopId;
  final String shopName;
  final String id;
  final String name;
  final double price;
  final double discount;
  final int quantity;
  final String imageCover;

  CheckoutProduct({
    required this.shopId,
    required this.shopName,
    required this.id,
    required this.name,
    required this.price,
    required this.discount,
    required this.quantity,
    required this.imageCover,
  });

  factory CheckoutProduct.fromJson(Map<String, dynamic> json) {
    return CheckoutProduct(
      shopId: json['shopId'],
      shopName: json['shopName'],
      id: json['_id'],
      name: json['name'],
      price: json['price'].toDouble(),
      discount: json['discount'].toDouble(),
      quantity: json['quantity'],
      imageCover: json['imageCover'],
    );
  }

  factory CheckoutProduct.empty() {
    return CheckoutProduct(
      shopId: '',
      shopName: '',
      id: '',
      name: '',
      price: 0.0,
      discount: 0.0,
      quantity: 0,
      imageCover: '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'shopId': shopId,
      'shopName': shopName,
      '_id': id,
      'name': name,
      'price': price,
      'discount': discount,
      'quantity': quantity,
      'imageCover': imageCover,
    };
  }
}
